framework/include/utils/NestedSolve.h

```diff
@@ -318,40 +420,290 @@ NestedSolve::nonlinear(V & guess, T compute)
   }
   auto r_square = r0_square;
 
-  // lambda to check for convergence and set the state accordingly
-  auto is_converged = [&]()
-  {
-    if (r_square < _absolute_tolerance_square)
-    {
-      _state = State::CONVERGED_ABS;
-      return true;
-    }
-    if (r_square / r0_square < _relative_tolerance_square)
-    {
-      _state = State::CONVERGED_REL;
-      return true;
-    }
-    return false;
-  };
-
   // perform non-linear iterations
   while (_n_iterations < _max_iterations)
   {
     // check convergence
-    if (_n_iterations >= _min_iterations && is_converged())
+    if (_n_iterations >= _min_iterations && isConverged(r0_square, r_square, /*acceptable=*/false))
       return;
 
     // solve and apply next increment
     linear(jacobian, delta, residual);
+
+    // Check if step size is smaller than the floating point tolerance
+    if (normSquare(delta) <= _x_tolerance_square)
```

Comment:
@ChaitanyaBhave-e this is not the same as the componentwise abs, but this method has overloads for every value type.